File: frontend/src/app/models/dynasty-event.ts
import type { Visibility } from './person';

export type DynastyEventType =
  | 'birth'
  | 'death'
  | 'coronation'
  | 'wedding'
  | 'war'
  | 'battle'
  | 'treaty'
  | 'other';

export const DYNASTY_EVENT_TYPES: readonly DynastyEventType[] = [
  'birth',
  'death',
  'coronation',
  'wedding',
  'war',
  'battle',
  'treaty',
  'other',
] as const;

/**
 * A dated happening, cross-linked to people by id rather than nested under
 * them. Stored in its own collection (`data/{world}-events.json`).
 */
export interface DynastyEvent {
  id: string;
  title: string;
  type: DynastyEventType;
  /** Day-of-month and month are optional precision on top of `startYear`, mirroring `Person.birthYear/Month/Day`. */
  startDay?: number;
  startMonth?: number;
  startYear: number;
  /**
   * Optional end of a spanning event (a reign, a war, a siege). Day-of-month
   * and month are optional precision on top of `endYear`, mirroring
   * `Person.deathYear/Month/Day` — unlike `start*`, none of these three are
   * required, since most events are a single dated happening, not a span.
   */
  endDay?: number;
  endMonth?: number;
  endYear?: number;
  era?: string;
  description?: string;
  relatedPersonIds: string[];
  /**
   * Which nation/house timelines this event appears on. Manually curated —
   * independent of `relatedPersonIds` overlap (see `TimelineDataService.filteredEvents`)
   * so a GM can pin an event onto a timeline it wouldn't otherwise qualify for.
   */
  nations: string[];
  houses: string[];
  /**
   * Same meaning as on `Person`: a secret war or hidden coronation drops out
   * of Player Preview the same way a secret person does.
   */
  visibility: Visibility;
}

File: tools/src/DynastyTools/Models/DynastyEvent.cs
namespace DynastyTools.Models;

/// <summary>
/// One dated happening, cross-linked to people by id. Mirrors the Angular
/// <c>DynastyEvent</c> interface.
/// </summary>
public sealed class DynastyEvent
{
    public string Id { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;

    /// <summary>
    /// Kept as a string so an unrecognised type is a validation finding rather
    /// than a deserialization crash. See <see cref="ValidTypes"/>.
    /// </summary>
    public string Type { get; set; } = "other";

    public int Year { get; set; }
    public string? Era { get; set; }
    public string? Description { get; set; }
    public List<string> RelatedPersonIds { get; set; } = [];
    public string? Nation { get; set; }
    public string? House { get; set; }
    public string Visibility { get; set; } = "known";

    public static readonly string[] ValidTypes =
    [
        "birth",
        "death",
        "coronation",
        "wedding",
        "war",
        "battle",
        "treaty",
        "other",
    ];
}

/// <summary>Shape of <c>data/{world}-events.json</c>.</summary>
public sealed class WorldEventsFile
{
    public List<DynastyEvent> Events { get; set; } = [];
}
